File: materialchecklist/src/main/java/com/dvdb/materialchecklist/util/LinksMovementMethod.java
package com.dvdb.materialchecklist.util;

import android.text.Layout;
import android.text.Selection;
import android.text.Spannable;
import android.text.method.ArrowKeyMovementMethod;
import android.text.method.MovementMethod;
import android.text.style.ClickableSpan;
import android.view.MotionEvent;
import android.widget.TextView;

public class LinksMovementMethod extends ArrowKeyMovementMethod {
    private static LinksMovementMethod mInstance = null;

    public static MovementMethod getInstance() {
        if (mInstance == null) {
            mInstance = new LinksMovementMethod();
        }
        return mInstance;
    }

    @Override
    public boolean onTouchEvent(TextView widget, Spannable buffer, MotionEvent event) {
        final int action = event.getAction();
        if (action == MotionEvent.ACTION_UP || action == MotionEvent.ACTION_DOWN) {
            int x = (int) event.getX();
            int y = (int) event.getY();

            x -= widget.getTotalPaddingLeft();
            y -= widget.getTotalPaddingTop();

            x += widget.getScrollX();
            y += widget.getScrollY();

            final Layout layout = widget.getLayout();
            int line = layout.getLineForVertical(y);
            int off = layout.getOffsetForHorizontal(line, x);

            final ClickableSpan[] link = buffer.getSpans(off, off, ClickableSpan.class);
            if (link.length != 0) {
                if (action == MotionEvent.ACTION_UP) {
                    link[0].onClick(widget);
                } else {
                    Selection.setSelection(buffer, buffer.getSpanStart(link[0]), buffer.getSpanEnd(link[0]));
                }

                return true;
            }
        }
        return super.onTouchEvent(widget, buffer, event);
    }
}

File: materialchecklist/src/main/java/com/dvdb/materialchecklist/util/LinksMovementMethod.kt
package com.dvdb.materialchecklist.util

import android.text.Selection
import android.text.Spannable
import android.text.method.ArrowKeyMovementMethod
import android.text.style.ClickableSpan
import android.view.MotionEvent
import android.widget.TextView

internal object LinksMovementMethod : ArrowKeyMovementMethod() {

    override fun onTouchEvent(
        widget: TextView,
        buffer: Spannable,
        event: MotionEvent
    ): Boolean {
        val action = event.action
        if (action == MotionEvent.ACTION_UP ||
            action == MotionEvent.ACTION_DOWN
        ) {
            val x = event.x.toInt() - widget.totalPaddingLeft + widget.scrollX
            val y = event.y.toInt() - widget.totalPaddingTop + widget.scrollY
            val line = widget.layout.getLineForVertical(y)
            val horizontalOffset = widget.layout.getOffsetForHorizontal(
                line,
                x.toFloat()
            )

            val link = buffer.getSpans(
                horizontalOffset,
                horizontalOffset,
                ClickableSpan::class.java
            )

            if (link.isNotEmpty()) {
                if (action == MotionEvent.ACTION_UP) {
                    link.first().onClick(widget)
                } else {
                    Selection.setSelection(
                        buffer,
                        buffer.getSpanStart(link.first()),
                        buffer.getSpanEnd(link.first())
                    )
                }
                return true
            }
        }
        return super.onTouchEvent(widget, buffer, event)
    }
}